pennylane/templates/state_preparations/mottonen.py

```diff
@@ -14,83 +14,107 @@
 r"""
 Contains the MottonenStatePreparation template.
 """
+from typing import Optional
+
 import numpy as np
 
 import pennylane as qml
 from pennylane.operation import Operation
+from pennylane.typing import TensorLike
 
 
-# pylint: disable=len-as-condition,arguments-out-of-order,consider-using-enumerate
 def gray_code(rank):
-    """Generates the Gray code of given rank.
+    """Generates the Gray code of given rank, as numeric output.
 
     Args:
         rank (int): rank of the Gray code (i.e. number of bits)
-    """
-
-    def gray_code_recurse(g, rank):
-        k = len(g)
-        if rank <= 0:
-            return
-
-        for i in range(k - 1, -1, -1):
-            char = "1" + g[i]
-            g.append(char)
-        for i in range(k - 1, -1, -1):
-            g[i] = "0" + g[i]
-
-        gray_code_recurse(g, rank - 1)
-
-    g = ["0", "1"]
-    gray_code_recurse(g, rank - 1)
-
-    return g
-
-
-def _matrix_M_entry(row, col):
-    """Returns one entry for the matrix that maps alpha to theta.
-
-    See Eq. (3) in `Möttönen et al. (2004) <https://arxiv.org/abs/quant-ph/0407010>`_.
-
-    Args:
-        row (int): one-based row number
-        col (int): one-based column number
 
     Returns:
-        (float): transformation matrix entry at given row and column
+        np.ndarray[int]: Array of ``2**rank`` integers that make up the Gray code.
     """
-    # (col >> 1) ^ col is the Gray code of col
-    b_and_g = row & ((col >> 1) ^ col)
-    sum_of_ones = 0
-    while b_and_g > 0:
-        if b_and_g & 0b1:
-            sum_of_ones += 1
+    g = np.array([0, 1])
+    for i in range(1, rank):
+        g = np.concatenate([g, g[::-1] + 2**i])
+    return g
 
-        b_and_g = b_and_g >> 1
 
-    return (-1) ** sum_of_ones
+_walsh_hadamard_matrix = np.array([[1, 1], [1, -1]]) / 2
+_cnot_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]]).reshape((2,) * 4)
 
 
-def compute_theta(alpha):
-    """Maps the angles alpha of the multi-controlled rotations decomposition of a uniformly controlled rotation
-     to the rotation angles used in the Gray code implementation.
+def compute_theta(alpha: TensorLike, num_qubits: Optional[int] = None):
+    r"""Maps the angles alpha of the multi-controlled rotations decomposition of a
+    uniformly controlled rotation to the rotation angles used in the Gray code implementation.
+    This function uses the fact that the transformation given by Eq. (3) in
+    `Möttönen et al. (2004) <https://arxiv.org/abs/quant-ph/0407010>`_ is equal to a Walsh-Hadamard
+    transform followed by some permutations, which can be expressed as a ladder of CNOT gates
+    applied to the angles, when interpreting them as a quantum state.
```

Comment:
What do you mean here in "when interpreting them as a quantum state."?

Reply:
Ah, I just wanted to make clear that we are "applying cnot gates" to something that is just an array of angles, and not a state vector. So "applying a gate" is somewhat wrong but it is what we do computationally :D